### PR TITLE
Add feature of signing with es25519 and sr25519

### PR DESCRIPTION
**Pull request by benxgao**

### CODE REVIEW

File: src/signature/mod.rs [hunk 1]
```diff
@@ -24,6 +29,51 @@ impl TryFrom<u8> for SignatureVersion {
     }
 }
 
+/// copied from wasm/wasm-crypto
+fn create_from_pair(pair: &[u8]) -> EdKeypair {
+    match EdKeypair::from_bytes(pair) {
+        Ok(pair) => return pair,
+        Err(_) => panic!("Provided pair is invalid."),
```
Comment (alexsednz) [inline src/signature/mod.rs:36]: I am against panicking unless it is for test code. I would suggest we use EdKeypair::from_bytes(pair) itself where we need to create a key from its pair and in that context handle the error.

File: src/signature/mod.rs [hunk 2]
```diff
@@ -24,6 +29,51 @@ impl TryFrom<u8> for SignatureVersion {
...
+    pair.extend_from_slice(pubkey);
+
+    create_from_pair(&pair)
+}
+
+/// copied from wasm/wasm-crypto
+fn create_public(pubkey: &[u8]) -> Ed25519Pub {
+    match Ed25519Pub::from_bytes(pubkey) {
+        Ok(pubkey) => return pubkey,
+        Err(_) => panic!("Provided public key is invalid."),
```
Comment (alexsednz) [inline src/signature/mod.rs:54]: Same here I'd prefer the direct use of Ed25519Pub::from_bytes(pubkey)

File: .rustfmt.toml [hunk 3]
```diff
@@ -1 +1 @@
-license_template_path = "LICENSE_TEMPLATE"
\ No newline at end of file
+# license_template_path = "LICENSE_TEMPLATE"
```
Comment (alexsednz) [inline .rustfmt.toml:1]: Why did you need this change?

Comment (benxgao) [reply .rustfmt.toml:1]: `cargo fmt` throws linting error about it, comment it for now.

File: src/signature/mod.rs [hunk 4]
```diff
@@ -39,6 +89,16 @@ pub fn verify_signature(
     }
 }
 
+/// Verify an ed25519 signature
+pub fn ed_verify(signature: &[u8], payload: &[u8], pubkey: &[u8]) -> bool {
+    let signature = match Ed25519Sig::from_bytes(signature) {
```
Comment (holygits) [inline src/signature/mod.rs:94]: It's good to return errors to the caller so the return type would be smth like: `Result<bool, &str>`
Ok(true) no error and sig ok
Ok(false) no error sig bad
Err() error happened, sig unknown

File: src/signature/mod.rs [hunk 5]
```diff
@@ -24,6 +29,51 @@ impl TryFrom<u8> for SignatureVersion {
...
+    create_from_parts(pubkey.as_bytes(), seed)
+}
+
+/// copied from wasm/wasm-crypto
+pub fn create_ed_keypair_from_seed(seed: &[u8]) -> Vec<u8> {
+    create_from_seed(seed).to_bytes().to_vec()
+}
+
+/// Sign an ed25519 signature
+pub fn ed_sign(keypair: EdKeypair, payload: &[u8]) -> Vec<u8> {
```
Comment (holygits) [inline src/signature/mod.rs:72]: function signature should be: `sign_ed25519(<private key>: &[u8], <payload>: &[u8]) -> Vec<u8>`
This will give the caller the most flexibility to provide the private key however they want. Otherwise it's unnecessarily restricted to use `EdKeypair` crate / struct only.

next step this function should turn into a method on doughnut structs.
to use like: `doughnut.sign_ed25519(<private_key>)`
A trait will help, then impl the trait for the Doughnut structs to sign and append the signature
e.g.
```
/// Doughnut signing methods
trait Signing {
    /// sign using Sr25519 method
    fn sign_sr25519(&mut self, private_key: &[u8]);
    /// sign using Ed25519 method
    fn sign_ed25519(&mut self, private_key: &[u8]);
}
```

File: src/signature/mod.rs [hunk 6]
```diff
@@ -24,6 +29,51 @@ impl TryFrom<u8> for SignatureVersion {
...
+/// copied from wasm/wasm-crypto
+fn create_from_pair(pair: &[u8]) -> EdKeypair {
+    match EdKeypair::from_bytes(pair) {
+        Ok(pair) => return pair,
+        Err(_) => panic!("Provided pair is invalid."),
+    }
+}
+
+/// copied from wasm/wasm-crypto
+fn create_from_parts(pubkey: &[u8], seckey: &[u8]) -> EdKeypair {
```
Comment (holygits) [inline src/signature/mod.rs:41]: too many `create_*` functions, hard to follow, would recommend inlining them all into one function for verify and one function for sign

**Comment by benxgao** [comment]
Thanks @holygits , those make sense, will refactor them based on your feedback

File: src/traits/mod.rs [hunk 7]
```diff
@@ -65,6 +65,14 @@ pub trait DoughnutApi {
     }
 }
 
+pub trait Signing {
```
Comment (benxgao) [inline src/traits/mod.rs:68]: @holygits What if here we have only one single entry for `fn sign(...args)`, and encapsulate various kind of signing methods, only if signing function always takes `publickey, secretkey, payload` as arguments?

cc @alexsednz @KarishmaBothara

File: src/traits/impls.rs [hunk 8]
```diff
@@ -46,6 +48,51 @@ impl DoughnutApi for () {
...
+    }
+
+    fn sign_sr25519(&self, public_key: &[u8], private_key: &[u8], payload: &[u8]) -> Result<(), VerifyError> {
+        sign_sr25519(&public_key, &private_key, &payload);
+    }
+}
+
+#[cfg(feature = "std")]
+#[allow(unreachable_patterns)]
+impl Signing for Doughnut {
```
Comment (benxgao) [inline src/traits/impls.rs:85]: @holygits Shall we implement signing for all above Doughnuts, `(), DoughnutV0, ParityDoughnutV0 `?

File: src/signatures/mod.rs [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        Err(_) => return false,
+    };
+
+    create_ed25519_public_key(public_key)
+        .verify(payload, &signature)
+        .is_ok()
+}
+
+/// Sign an sr25519 signature
+pub fn sign_sr25519(public_key: &[u8], private_key: &[u8], payload: &[u8]) -> Vec<u8> {
```
Comment (benxgao) [inline src/signatures/mod.rs:63]: @holygits I relocated new signing/verifying functions into a separated module, and will merge it with the existing verifying method if necessary before final review/PR merge, is it ok?

File: src/signatures/mod.rs [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+            CONTEXT_ID,
+            payload,
+            &Sr25519PublicKey::from_bytes(public_key).unwrap(),
+        )
+        .to_bytes()
+        .to_vec()
+}
+
+/// Verify an sr25519 signature
+pub fn verify_sr25519(signature: &[u8], payload: &[u8], public_key: &[u8]) -> bool {
```
Comment (benxgao) [inline src/signatures/mod.rs:76]: @holygits I relocated new signing/verifying functions into a separated module called `mod signatures`, and will merge with current existing `verify functions` inside `mod signature` if necessary before final review/PR merge, is it ok?

File: src/traits/mod.rs (hunk 7 above)
Comment (benxgao) [reply src/traits/mod.rs:68]: Umm... maybe it would be better to expose determined api interfaces like
```js
sign_sr25519(pubkey, scrtkey, payload);
// or
sign_ed25519(pubkey, scrtkey, payload);
```
rather than
```js
sign(pubkey, scrtkey, payload, 'sr25519');
```

File: src/signatures/mod.rs (hunk 10 above)
Comment (holygits) [reply src/signatures/mod.rs:76]: sounds good

File: src/traits/impls.rs [hunk 11]
```diff
@@ -46,6 +48,51 @@ impl DoughnutApi for () {
...
+        sign_ed25519(&public_key, &private_key, &payload);
+    }
+
+    fn sign_sr25519(&self, public_key: &[u8], private_key: &[u8], payload: &[u8]) -> Result<(), VerifyError> {
+        sign_sr25519(&public_key, &private_key, &payload);
+    }
+}
+
+#[cfg(feature = "std")]
+impl<'a> Signing for DoughnutV0<'a> {
```
Comment (holygits) [inline src/traits/impls.rs:62]: Can ignore this one, it will be removed in #50

File: src/traits/mod.rs (hunk 7 above)
Comment (holygits) [reply src/traits/mod.rs:68]: don't need `payload` arg anymore, can use the doughnut directly `self.payload()`.
not sure we need to pass `public_key` at all, `from_seed_*` functions will be enough to create the keypair

File: src/signatures/mod.rs [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    };
+
+    create_ed25519_public_key(public_key)
+        .verify(payload, &signature)
+        .is_ok()
+}
+
+/// Sign an sr25519 signature
+pub fn sign_sr25519(public_key: &[u8], private_key: &[u8], payload: &[u8]) -> Vec<u8> {
+    Sr25519SecretKey::from_ed25519_bytes(private_key)
```
Comment (holygits) [inline src/signatures/mod.rs:64]: is this the correct (from_ed25519)?

File: src/traits/impls.rs (hunk 8 above)
Comment (holygits) [reply src/traits/impls.rs:85]: Can start with `ParityDoughnutV0`, an impl for `()` will only be useful for tests. It's not used at the moment so would leave it out

File: src/traits/mod.rs [hunk 13]
```diff
@@ -65,6 +65,14 @@ pub trait DoughnutApi {
...
+    /// sign using Sr25519 method
+    fn sign_sr25519(&mut self, public_key: &[u8], private_key: &[u8], payload: &[u8]);
```
Comment (holygits) [inline src/traits/mod.rs:70]: `fn sign_sr25519(&mut self, private_key: &[u8]);`

File: src/signatures/mod.rs (hunk 12 above)
Comment (benxgao) [reply src/signatures/mod.rs:64]: Ref here: https://github.com/polkadot-js/wasm/blob/master/packages/wasm-crypto/src/sr25519.rs#L115

File: src/signature/mod.rs [hunk 14]
```diff
@@ -24,6 +33,49 @@ impl TryFrom<u8> for SignatureVersion {
...
+        Err(_) => Err(SigningError::InvalidEd25519Key),
+    }
+}
+
+/// Sign an ed25519 signature
+pub fn sign_ed25519(
+    public_key: &[u8],
+    secret_key: &[u8],
+    payload: &[u8],
+) -> SigningResult<Vec<u8>> {
```
Comment (benxgao) [inline src/signature/mod.rs:53]: Do I need to check format of `public_key` and `secret_key` here?

File: src/signature/mod.rs [hunk 15]
```diff
@@ -72,19 +126,94 @@ mod test {
...
+
+        create_ed25519_keypair_from_keys(pubkey.as_bytes(), seed).unwrap()
+    }
+
+    /// copied from wasm/wasm-crypto
+    fn create_ed25519_keypair_from_seed(seed: &[u8]) -> Vec<u8> {
+        create_from_seed(seed).to_bytes().to_vec()
+    }
+
+    fn create_sr25519_keypair_from_seed(seed: &[u8]) -> Vec<u8> {
```
Comment (holygits) [inline src/signature/mod.rs:161]: This is where to check the key format errors
`schnorrkel::MiniSecretKey::from_bytes(seed).unwrap()` will fail if the format is bad.
can return the error here
also not good to use `unwrap` unless you know it will succeed, otherwise use `expect("some message")` if you really want to panic. Best thing is to return the error in libraries

File: src/signature/mod.rs [hunk 16]
```diff
@@ -72,19 +126,94 @@ mod test {
...
+    fn create_from_seed(seed: &[u8]) -> Ed25519Keypair {
+        let seckey = ed25519_dalek::SecretKey::from_bytes(seed).unwrap();
+        let pubkey: Ed25519PublicKey = (&seckey).into();
+
+        create_ed25519_keypair_from_keys(pubkey.as_bytes(), seed).unwrap()
+    }
+
+    /// copied from wasm/wasm-crypto
+    fn create_ed25519_keypair_from_seed(seed: &[u8]) -> Vec<u8> {
+        create_from_seed(seed).to_bytes().to_vec()
```
Comment (holygits) [inline src/signature/mod.rs:158]: no need to have this function call another function, just use one function

File: src/traits/impls.rs [hunk 17]
```diff
@@ -138,24 +166,25 @@ mod test {
         let header: Vec<u8> = vec![0, 0, 3];
         let issuer = keypair.public.to_bytes().to_vec();
         let holder = vec![0x15; 32];
-
         let payload: Vec<u8> = [header, issuer, holder, test_domain_data()].concat();
-        let signature = keypair.sign(context.bytes(&payload));
-
-        let encoded: Vec<u8> = [payload, signature.to_bytes().to_vec()].concat();
+        let signature_bytes = keypair.sign(context.bytes(&payload)).to_bytes().to_vec();
```
Comment (holygits) [inline src/traits/impls.rs:170]: This test is currently signing using some helper functions.
Goal is to test whether the new signing functions are working:
```
// it's unsigned
let signature  = H512::default();
let encoded: Vec<u8> = [payload, signature.to_bytes().to_vec()].concat();

let mut doughnut: ParityDoughnutV0 =
            Decode::decode(&mut &encoded[..]).expect("It is a valid doughnut");

// sign the doughnut
let secret_key = keypair.secret.to_ed25519_bytes();
let _ = doughnut.sign_sr25519(&secret_key).expect("it signed ok");

// verify the doughnut signature
assert_eq!(doughnut.verify(), Ok(()));
```

File: src/signature/mod.rs [hunk 18]
```diff
@@ -24,6 +33,49 @@ impl TryFrom<u8> for SignatureVersion {
     }
 }
 
+fn create_ed25519_keypair_from_keys(pubkey: &[u8], seckey: &[u8]) -> SigningResult<Ed25519Keypair> {
+    let mut pair = vec![];
```
Comment (holygits) [inline src/signature/mod.rs:37]:
```
let pair: Vec<u8> = [seckey.to_vec(), pubkey.to_vec()].concat();
Ed25519Keypair::from_bytes(&pair).map_err(|_| SigningError::InvalidEd25519Key)
```

File: src/signature/mod.rs [hunk 19]
```diff
@@ -72,19 +126,94 @@ mod test {
...
+    #[test]
+    fn can_sign_sr25519() {
+        let seed = generate_random_seed();
+        let keypair = create_sr25519_keypair_from_seed(seed.as_slice());
+        let private = &keypair[0..SR25519_SECRET_KEY_LENGTH];
+        let public = &keypair[SR25519_SECRET_KEY_LENGTH..SR25519_KEYPAIR_LENGTH];
+        let payload = b"this is a payload";
+        let signature = sign_sr25519(public, private, payload).unwrap();
+
+        assert!(signature.len() == SR25519_SIGNATURE_LENGTH);
```
Comment (holygits) [inline src/signature/mod.rs:216]: should verify signatures are correct here also

File: src/traits/impls.rs [hunk 20]
```diff
@@ -18,7 +22,8 @@ use crate::{
 impl DoughnutApi for () {
     type PublicKey = [u8; 32];
     type Timestamp = u32;
-    type Signature = ();
+    type Signature = H512;
```
Comment (holygits) [inline src/traits/impls.rs:25]: Would prefer this was left as `()` or a really basic type like: `[u8 ;64]]`.
It's not version 0 specifc, so not `H512`

**Comment by benxgao** [comment]
@holygits Refactoring based on review has been put in this commit, could you plz take a look once you are available, thanks. https://github.com/cennznet/doughnut-rs/pull/52/commits/832f47aeabda4ecf16b5a8be5b4865daee829c36

File: src/traits/impls.rs [hunk 21]
```diff
@@ -46,6 +51,23 @@ impl DoughnutApi for () {
     }
 }
 
+#[cfg(feature = "std")]
+impl Signing for ParityDoughnutV0 {
+    fn sign_ed25519(&mut self, secret_key: &[u8]) -> Result<H512, SigningError> {
```
[the remark this answers is not in the document]
Comment (holygits) [reply src/traits/impls.rs:56]: For convenience, set signature here would be good.
More advanced user can always make mutable doughnut and sign and set the signature externally.

File: src/signature/mod.rs [hunk 22]
```diff
@@ -24,6 +30,45 @@ impl TryFrom<u8> for SignatureVersion {
...
+        .map(|keypair| keypair.sign(payload).to_bytes().to_vec())
+}
+
+/// Sign an sr25519 signature
+pub fn sign_sr25519(
+    public_key: &[u8],
+    secret_key: &[u8],
+    payload: &[u8],
+) -> Result<Vec<u8>, SigningError> {
+    Sr25519SecretKey::from_ed25519_bytes(secret_key)
```
Comment (holygits) [inline src/signature/mod.rs:58]: Need to check the pub key format also instead of `unwrap` incase it fails.
This style may help
```rust

let secret_key_ = Sr25519SecretKey::from_ed25519_bytes(secret_key).map_err(|_| InvalidSr25519SecretKey)?;
let public_key_ = Sr25519PublicKey::from_bytes(public_key).map_err(InvalidSr25519PublicKey)?;

secret_key_.sign_simple(CONTEXT_ID, payload, &public_key_)
  .to_bytes()
  .to_vec()
```

File: src/traits/impls.rs [hunk 23]
```diff
@@ -46,6 +51,23 @@ impl DoughnutApi for () {
...
+        match sign_ed25519(&self.issuer(), secret_key, &self.payload()) {
+            Ok(signed_signature) => Ok(H512::from_slice(&signed_signature.as_slice())),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn sign_sr25519(&mut self, secret_key: &[u8]) -> Result<H512, SigningError> {
+        match sign_sr25519(&self.issuer(), secret_key, &self.payload()) {
+            Ok(signed_signature) => Ok(H512::from_slice(&signed_signature.as_slice())),
```
Comment (holygits) [inline src/traits/impls.rs:65]: Can we change this to return `Vec<u8>` since it's a simpler type for clients (i.e. they won't need to import H512).
`H512` is an unfortunate type we use for compatibility in plug/substrate.

File: src/signature/mod.rs (hunk 22 above)
Comment (benxgao) [reply src/signature/mod.rs:58]: @holygits In Rust, style of `let xxxx_ = xxx;` is our suggested naming convention for local variable? How about private functions, like `fn _xx(() {...}`?

File: src/traits/impls.rs (hunk 23 above)
Comment (benxgao) [reply src/traits/impls.rs:65]: It seems signature is defined as H512 in ParityDoughnutV0 - https://github.com/cennznet/doughnut-rs/blob/enable-signing/src/v0/parity.rs#L228

Comment (benxgao) [reply src/traits/impls.rs:65]: @holygits Is there a type like `any`?

File: src/signature/mod.rs (hunk 22 above)
Comment (holygits) [reply src/signature/mod.rs]: It's not convention couldn't think of a better name 🤷‍♀️

File: src/traits/impls.rs [hunk 24]
```diff
@@ -42,6 +51,27 @@ impl DoughnutApi for () {
     }
 }
 
+#[cfg(feature = "std")]
+impl Signing for DoughnutV0 {
+    fn sign_ed25519(&mut self, secret_key: &[u8]) -> Result<H512, SigningError> {
+        sign_ed25519(&self.issuer(), secret_key, &self.payload())
```
Comment (holygits) [inline src/traits/impls.rs:57]: Will need to set the signature version in these functions also
`self.signature_version = SignatureVersion::Ed25519 as u8;`
Could add some assertion to check signature version in tests also

File: src/traits/impls.rs [hunk 25]
```diff
@@ -122,58 +152,121 @@ mod test {
...
+
+        assert_eq!(
+            doughnut.sign_ed25519(&secret_key),
+            Err(SigningError::InvalidEd25519Key)
+        );
+    }
+
+    #[test]
+    fn sr25519_signed_doughnut_v0_has_invalid_signature() {
+        let keypair = generate_sr25519_keypair();
```
[the remark this answers is not in the document]
Comment (holygits) [reply src/traits/impls.rs:253]: nvm I see now :)

File: src/traits/impls.rs [hunk 26]
```diff
@@ -122,58 +152,121 @@ mod test {
...
+
+        // Signature version = 0
+        // has not before (b0) and 2 domains (b1..7)
+        let header: Vec<u8> = vec![0, 0, 3];
         let issuer = keypair.public.to_bytes().to_vec();
         let holder = vec![0x15; 32];
 
         let payload: Vec<u8> = [header, issuer, holder, test_domain_data()].concat();
-        let signature = keypair.sign(&payload);
+        let invalid_signature = keypair_invalid.sign(context.bytes(&payload));
```
[the remark this answers is not in the document]
Comment (holygits) [reply src/traits/impls.rs:264]: nvm here, using invalid key not invalid payload
